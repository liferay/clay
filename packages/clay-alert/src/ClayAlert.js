import './ClayAlertBase';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import templates from './ClayAlert.soy.js';

/**
 * Metal Clay Alert component.
 */
class ClayAlert extends Component {
	/**
	 * @inheritDoc
	 */
	attached() {
		this.addListener('hide', this._defaultHideAlert, true);
	}

	/**
	 * Hides the alert and destroy it if proceed.
	 * @private
	 */
	_defaultHideAlert() {
		if (this.destroyOnHide) {
			this.dispose();
		}
	}

	/**
	 * Continues the propagation of the hide event
	 * @return {Boolean} If the event has been prevented or not.
	 */
	_handleHide() {
		return !this.emit('hide');
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayAlert.STATE = {
	/**
	 * Flag to indicate if the alert is closeable.
	 * @instance
	 * @memberof ClayAlert
	 * @type {?bool}
	 * @default false
	 */
	closeable: Config.bool().value(false),

	/**
	 * Flag to indicate if the alert should be destroyen when close.
	 * @instance
	 * @memberof ClayAlert
	 * @type {?bool}
	 * @default false
	 */
	destroyOnHide: Config.bool().value(false),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayAlert
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayAlert
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * The message of alert
	 * @instance
	 * @memberof ClayAlert
	 * @type {!html|string}
	 */
	message: Config.any().required(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayAlert
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),

	/**
	 * The style of alert
	 * @instance
	 * @memberof ClayAlert
	 * @type {?string}
	 * @default info
	 */
	style: Config.oneOf(['danger', 'info', 'success', 'warning']).value('info'),

	/**
	 * The title of alert
	 * @instance
	 * @memberof ClayAlert
	 * @type {!string}
	 * @default undefined
	 */
	title: Config.string().required(),
};

defineWebComponent('clay-alert', ClayAlert);

Soy.register(ClayAlert, templates);

export {ClayAlert};
export default ClayAlert;
