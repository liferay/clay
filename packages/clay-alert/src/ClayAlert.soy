{namespace ClayAlert}

/**
 * This renders the component's whole content.
 */
{template .render}
    {@param message: html|string}
    {@param spritemap: string}
    {@param title: string}
    {@param? _handleHide: any}
    {@param? closeable: bool}
    {@param? elementClasses: string}
    {@param? id: string}
    {@param? style: string}

    {call ClayAlertBase.render}
        {param closeable: $closeable /}
        {param elementClasses: $elementClasses /}
        {param events: ['hide': $_handleHide ] /}
        {param id: $id /}
        {param message: $message /}
        {param ref: 'alertBase' /}
        {param spritemap: $spritemap /}
        {param style: $style /}
        {param title: $title /}
    {/call}
{/template}
