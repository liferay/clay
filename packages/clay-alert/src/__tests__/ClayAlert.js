import ClayAlert from '../ClayAlert';

let consoleErrorReference;
let alert;
const spritemap = '../node_modules/clay-css/lib/images/icons/icons.svg';

/**
 * Stubs console.error
 */
function mockConsoleError() {
	console.error = () => {};
}

describe('ClayAlert', function() {
	beforeEach(() => {
		consoleErrorReference = console.error;
	});

	afterEach(() => {
		if (alert) {
			alert.dispose();
		}

		console.error = consoleErrorReference;
	});

	it('should render default markup', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			title: 'info',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render an alert with classes', function() {
		alert = new ClayAlert({
			elementClasses: 'my-custom-class',
			message: 'message',
			spritemap: spritemap,
			title: 'info',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render an alert with id', function() {
		alert = new ClayAlert({
			id: 'myId',
			message: 'message',
			spritemap: spritemap,
			title: 'info',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render a "danger" style alert', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			style: 'danger',
			title: 'Danger!',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render an "info" style alert', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			style: 'info',
			title: 'Info',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render a "success" style alert', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			style: 'success',
			title: 'Success!',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render a "warning" style alert', function() {
		alert = new ClayAlert({
			message: 'message',
			spritemap: spritemap,
			style: 'warning',
			title: 'Warning!',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should render a closeable alert', function() {
		alert = new ClayAlert({
			closeable: true,
			message: 'message',
			spritemap: spritemap,
			style: 'warning',
			title: 'Warning!',
		});

		expect(alert).toMatchSnapshot();
	});

	it('should close a closeable alert', function() {
		alert = new ClayAlert({
			closeable: true,
			message: 'message',
			spritemap: spritemap,
			title: 'Success!',
		});

		alert.element.querySelector('button.close').click();

		expect(alert._visible).toBeFalsy();
	});

	it('should close and destroy an alert', function() {
		alert = new ClayAlert({
			closeable: true,
			destroyOnHide: true,
			message: 'message',
			spritemap: spritemap,
			title: 'Success!',
		});

		alert.element.querySelector('button.close').click();

		expect(alert._visible).toBeFalsy();
		expect(alert.disposed_).toBeTruthy();
	});

	it('should fail when no message is passed', function() {
		mockConsoleError();

		expect(() => {
			alert = new ClayAlert({
				spritemap: spritemap,
				title: 'title',
			});
		}).toThrow();
	});

	it('should fail when no spritemap is passed', function() {
		mockConsoleError();

		expect(() => {
			alert = new ClayAlert({
				message: 'message',
				title: 'title',
			});
		}).toThrow();
	});

	it('should fail when no title is passed', function() {
		mockConsoleError();

		expect(() => {
			alert = new ClayAlert({
				message: 'message',
				spritemap: spritemap,
			});
		}).toThrow();
	});
});
