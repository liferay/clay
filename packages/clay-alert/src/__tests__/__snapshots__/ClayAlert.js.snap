// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayAlert should render a "danger" style alert 1`] = `
<div class="alert alert-danger fade show" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-exclamation-full" focusable="false">
      <title>exclamation-full</title>
      <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#exclamation-full"></use>
    </svg>
  </span>
  <strong class="lead">Danger!</strong>message</div>
`;

exports[`ClayAlert should render a "success" style alert 1`] = `
<div class="alert alert-success fade show" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-check-circle-full" focusable="false">
      <title>check-circle-full</title>
      <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#check-circle-full"></use>
    </svg>
  </span>
  <strong class="lead">Success!</strong>message</div>
`;

exports[`ClayAlert should render a "warning" style alert 1`] = `
<div class="alert alert-warning fade show" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-warning-full" focusable="false">
      <title>warning-full</title>
      <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#warning-full"></use>
    </svg>
  </span>
  <strong class="lead">Warning!</strong>message</div>
`;

exports[`ClayAlert should render a closeable alert 1`] = `
<div class="alert alert-warning fade alert-dismissible show" role="alert">
  <button class="btn close btn-unstyled" aria-label="close" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-times" focusable="false">
      <title>times</title>
      <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#times"></use>
    </svg>
  </button>
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-warning-full" focusable="false">
      <title>warning-full</title>
      <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#warning-full"></use>
    </svg>
  </span>
  <strong class="lead">Warning!</strong>message</div>
`;

exports[`ClayAlert should render an "info" style alert 1`] = `
<div class="alert alert-info fade show" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle" focusable="false">
      <title>info-circle</title>
      <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#info-circle"></use>
    </svg>
  </span>
  <strong class="lead">Info</strong>message</div>
`;

exports[`ClayAlert should render an alert with classes 1`] = `
<div class="alert alert-info fade my-custom-class show" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle" focusable="false">
      <title>info-circle</title>
      <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#info-circle"></use>
    </svg>
  </span>
  <strong class="lead">info</strong>message</div>
`;

exports[`ClayAlert should render an alert with id 1`] = `
<div class="alert alert-info fade show" id="myId" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle" focusable="false">
      <title>info-circle</title>
      <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#info-circle"></use>
    </svg>
  </span>
  <strong class="lead">info</strong>message</div>
`;

exports[`ClayAlert should render default markup 1`] = `
<div class="alert alert-info fade show" role="alert">
  <span class="alert-indicator">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-info-circle" focusable="false">
      <title>info-circle</title>
      <use xlink:href="../node_modules/clay-css/lib/images/icons/icons.svg#info-circle"></use>
    </svg>
  </span>
  <strong class="lead">info</strong>message</div>
`;
